### DISCUSSION TITLE
Question about the TextPredictor (no categorical MLP by default)

### DISCUSSION THREAD

**Discussion by Qiaochu-Song** [autogluon/autogluon#1774] [Q&A] [answered]
Hi,

First thank you so much for the powerful Autogluon! 
I have one quick question: it seems that in the current version, TextPredictor pytorch backend will not have categorical MLP by default, which is different from the previous MXNet backend (and your paper that proposed for multimodal model "fuse-late multimodal net"). I am curious about why the default is changed in this way. Does it give better performance than using categorical MLP? 

Thanks!

**Comment by sxjscience** [comment, marked as answer]
It has categorical MLP by default: https://github.com/awslabs/autogluon/blob/1828cbc9e7f01f42ee559f1f11b57b53d0100e50/text/src/autogluon/text/automm/configs/model/fusion_mlp_image_text_tabular.yaml#L3

**Reply by Qiaochu-Song** [reply]
Thanks for your reply! I see the auto_mm does have the categorical MLP by default, but according to the [column type inference and data preprocessing setting](https://github.com/awslabs/autogluon/blob/0.4.1/text/src/autogluon/text/automm/configs/data/default.yaml#L8), autogluon will convert categorical fields to text by default, so that no categorical MLP will be used. 
I have also tested with your text-tabular benchmark dataset, and seems there will be no categorical MLP even if there are categorical columns. 
Previously I have created a PR for this (which is closed now): https://github.com/awslabs/autogluon/pull/1751. But I found directly setting categorical to text conversion to false will fail the emoji test.

**Reply by sxjscience** [reply]
Oh, I see your point. Yeah, this is due to another round of benchmark which figured out that converting categorical to text will give us better performance. (The root cause may be that some categorical columns are actually text-like).

However, we are still doing more experiments in improving the type inference logic. Happy to discuss with you more on that.

**Reply by sxjscience** [reply]
The ideal scenario is to extract some column-level features to determine if it should be treated as category / text. But it requires some profiling on the benchmark datasets.

We also recently added FT-Transformer (https://github.com/awslabs/autogluon/blob/1828cbc9e7f01f42ee559f1f11b57b53d0100e50/text/src/autogluon/text/automm/configs/model/fusion_mlp_image_text_tabular.yaml#L17, https://github.com/awslabs/autogluon/blob/1828cbc9e7f01f42ee559f1f11b57b53d0100e50/text/src/autogluon/text/automm/configs/model/fusion_mlp_image_text_tabular.yaml#L107), and noticed that it can sometimes give us better performance. But we are still trying to figure out a consistent solution.

**Reply by Qiaochu-Song** [reply]
Thanks for your reply and it would be great to discussion more:) 
One interesting thing I observed is that, when a dataset has many categorical columns and the column names makes sense as texts, by combining column names (with some preprocessing, like split("_") for most kaggle datasets) with the categorical content as text can give better performance. 
The dataset I have tried on is "airbnb", which has 37 categorical columns. the score raised from 0.407 to 0.488 (AG v0.4.1 on one gpu).

**Reply by sxjscience** [reply]
Do you think it makes sense to make such feature available in AutoGluon by default?

In fact, column names do contain useful information and can be treated as additional information provided by the customer. We are also discussing about asking the users to provide some higher-level descriptions of the task in hand (@Raldir and I have talked about it).

Would you like to join our slack channel for an in-depth discussion?

**Reply by Qiaochu-Song** [reply]
From my point of view, I think setting this by default could be helpful, but I am also curious whether we can tell if the column names are valid texts, so that we can avoid too many` <UNK>` that might squeeze useful words out of the max_text_len=512 (especially when there are many categorical columns). 
I am more than happy to join your slack channel!

**Reply by sxjscience** [reply]
Sounds good (One tiny point is that these won't be `<UNK>` due to the BPE algorithm). 

My email address is xshiab AT connect DOT ust DOT hk. We can connect there and I will add you to the discussion channel.

**Reply by sxjscience** [reply]
In fact, we can adopt the "prompting" idea. We construct a prompt based on  column_name and column_value so that we can combine the categorical features and text features as a single prompt message.

We can implement a bunch of templates for constructing prompts and profile the performance on the multimodal dataframe.

In addition, for detecting whether the column name is meaningful, we can assign a score to the column name via a pretrained language model.

**Reply by Qiaochu-Song** [reply]
Wow, I have also been thinking about last Friday,  borrowing the idea of prompting and templates! 
The column name score you proposed also sounds great. 
P.S. thanks for pointing out the BPE thing!
